### ISSUE TITLE
Support the custom TODO patterns

### ISSUE THREAD

**Issue by tukusejssirs** [BashSupport-Pro/bashsupport-pro#186]
JetBrain IDEs support TODOs (three levels). It also supports to define custom TODO patterns and colouring them.

![ide_settings_todo](https://github.com/user-attachments/assets/b9746ea5-60d8-4f76-81ab-de8893aa4e45)

Now, when your plugin is not installed, they all work as expected, however, when it is, the TODO comments are in the default comment colour instead of the defined TODO colour.

It would be nice if your plugin supports these custom TODO colours. :wink:

**Comment by jansorg** [comment]
Thanks! I wasn't even aware of custom colors. I'll see what I can do about this...

**Comment by jansorg** [comment]
@tukusejssirs Which IDE and version are you using?
Could you a pattern and its settings, which doesn't work for you?
I was unable to reproduce this with 2024.3 eap. 
![image](https://github.com/user-attachments/assets/ac1e42d3-ade0-4d65-b849-45455ed63c53)

**Comment by tukusejssirs** [comment]
I use RustRover 2024.2.4 and WebStorm 2024.2.4.

Note: Screenshots bellow were created using a plugin called [Easy Code Screenshots](https://plugins.jetbrains.com/plugin/20766-easy-code-screenshots), but the screenshots use exactly the same syntax hightlighting as in my IDEs.

When using `Shell Script (BashSupport Pro)` language:

![image](https://github.com/user-attachments/assets/04f84ea4-15d4-4c7c-9e06-75246d507168)

When using the builtin `Shell Script` language:

![image](https://github.com/user-attachments/assets/9c947251-eb73-4072-bb82-44b1065f1817)

Therefore, basically, I have noticed that:

- the TODO patterns highlighting does not work (at least) when there is another comment after the pattern comment (it does not matter if the subsequent comment is a `shellcheck` comment, regular comment, another pattern comment);
- only the last comment line is highlighted according to the pattern rules.

> Could you a pattern and its settings, which doesn't work for you?

I could give you the patterns in text (but they are all in the screenshot of the OP), however, I believe you should be able to reproduce the issue based on what I have described above in this comment.

Basically, it does not work with any pattern, the actual pattern does not work. What matters is multiple subsequent comment lines, esp multiple consequent pattern comment lines.

**Comment by jansorg** [comment]
Thanks, that was helpful.
I've tracked this down to the trailing `$` in your patterns. At least I think it's the or one of the causes.

`$` is end of input.
JetBrains Shell uses a very generic ToDo indexing (which is slow) and as far as I can tell `$` is matching the end of a line in this context. That's why every end-of-line is considered for a `$`.

But in most other supported languages of the IDE (I tested XML, Java, BashSupport Pro) the `$` indicates the end of the complete comment. That's why only the last line was highlighted in your examples

Could you try this without the `$`, please?
Would it be possible to keep it like that or is there a reason to use the `$`?

With this content:
```bash
#!/bin/bash

# Note: Another note
# Not a work item
# Note: Another note

echo
```

The custom pattern `\bnote\b.*$` is only matching the last line:
![image](https://github.com/user-attachments/assets/278ce30b-dd0c-4f7c-81d8-7f5c5d582714)

But without the `$` (thus `\bnote\b.*`)  all notes are matched:
![image](https://github.com/user-attachments/assets/dc2d3ded-9428-4b9a-b595-0998bb177a3d)

**Comment by tukusejssirs** [comment]
Sorry, I’ve somehow missed your latest comment.

You are right that the issue is in the regex, however the dollar sign is not the only issue, as it only highlights the TODO word, not the comment itself. This change makes the highlighting the same in the builtin `Shell Script` and `Shell Script (BashSupport Pro)`.

![ide_settings_todo_without_dollar](https://github.com/user-attachments/assets/16722dad-5ef3-4af1-80d5-a0b7a41a0eef)

I am not sure why, as `TODO(|: .*)$` should match either `TODO$` or `TODO: .*$`.

I think you are using totally different regex library than one used by the IDE.

Note that the TODO patterns are used in any file type, not only in Bash, therefore, I would like BasSupport Pro regex matching in TODO patterns behave the same way as everywhere else.

Anyway, here are my patterns:

- `(?<=(/?[/*]|@remarks|#|rem|REM|::|--|;) |\[//\]: # \()TODO(|: .*)$`;
- `/\*\n[\t ]*\* TODO: [^\n]+:\n[\s\S]*\*/`;
- `(?<=(/?[/*]|@remarks|#|rem|REM|::|--|;) |\[//\]: # \()FIXME(|: .*)$`;
- `/\*\n[\t ]*\* FIXME: [^\n]+:\n[\s\S]*\*/`;
- `(?<=(/?[/*]|@remarks|#|rem|REM|::|--|;) |\[//\]: # \()Note(|: .*)$`;
- `/\*\n[\t ]*\* Note: [^\n]+:\n[\s\S]*\*/`;
- `(?<=\* )@todo .*`;
- `(?<=\* )@fixme .*`;
- `(?<=\* )@remarks .*`.

I don’t say they are perfect, I have quickly hacked them together. They work most of the time though which is good enough for me.

**Comment by jansorg** [comment]
@tukusejssirs Thank you for providing the patterns and the feedback. I appreciate your time to report and debug this.

If my comment doesn't help, I'd appreciate a quick video call to debug this in your environment, if you don't mind. In this case, please write to mail@bashsupport.com.

I've checked the first pattern in detail: `(?<=(/?[/*]|@remarks|#|rem|REM|::|--|;) |\[//\]: # \()TODO(|: .*)$`

In summary, when the pattern is fixed to work for Java multi-line comments it's also working in BashSupport Pro's comments. I think you need to change your patterns (i.e. the `$` handling) to make it work in all language comments, including BashSupport Pro.

Regarding RegExp library: The IDE is abstracting most of the ToDo pattern matching and therefore BashSupport Pro is using the same library (i.e. the JVM's implementation) for ToDo matching.

### Java ToDo Highlighting

#### Original Pattern

With Java, this is only providing matching in single-line-comment. But with .java multi-line comments, it's the same as in BashSupport Pro. `$` in a pattern matches "end of input". In Java single-line comments it's the end of a line, in Java multi-line comnents it's "end of the last line".
![image](https://github.com/user-attachments/assets/13ca8444-3ceb-497f-8ebf-2b0d09e36ecc)

#### Pattern without trailing `$` 

If you remove the trailing `$`, then you get this for the Java file. As you can see, only the `TODO:` part is highlighted, but in the 2nd and  3rd comments all items and not just the last item of the 3rd.
![image](https://github.com/user-attachments/assets/c5376168-99dc-47d7-a048-869b3772b4ae)

#### Fixed Pattern without trailing `$` 

In the pattern the last part `TODO(|: .*)` would first attempt to match `TODO` and then `TODO: .*`. When the trailing `$` was present, the whole match was forced to continue until the end of input, which made it highlight the last line of the 3rd comment.
Fixing this part of the pattern as `TODO(: .*)?` or `TODO(: .*|)` makes it highlight all of the pattern:
![image](https://github.com/user-attachments/assets/808654a4-6738-4fc7-a2ad-474f315e7bed)

### BashSupport Pro ToDo Highlighting

Multiple lines of a BashSupport Pro comments are internally treated as a single element, i.e. it's similar to Java's multi-line comment and not like a Java single-line comment.
![image](https://github.com/user-attachments/assets/421deb15-87ac-4096-9511-5667e17d5464)

JetBrains Shell uses a slow implementation, which isn't used by most "serious" language implementations of JetBrains. That's why its highlighting is showing different symptoms compared to BashSupport Pro.

**Comment by jansorg** [comment]
I'm closing this, hoping that my last comment helped to resolve the problem.
Please let me know if your problem persists.
Thank you for reporting!

**Comment by tukusejssirs** [comment]
@jansorg, sorry for late reply. I am busy with other stuff.

---

> multi-line comments

I don't use multiline comments, as I prefer soft wrapping to hard wrapping, therefore I have intentionally disabled `Treat indented text on the following lines as part of the same TODO` so that when the line following a comment matching a TODO pattern has 3-4 extra spaces after the comment character, I don't want it to be syntactically highlighted as a TODO comment.

However, what I want (and what works in the built-in `Shell Script`) is the possibility of having multiple consecutive comments matching TODO patterns _line by line_.

> Java

Anyway, I don't code in Java nor Kotlin. :wink: I do code in JavaScript/TypeScript though. :wink:

> In Java single-line comments it's the end of a line, in Java multi-line comnents it's "end of the last line".

I want to treat even multi-line comments line by line.

> Fixed Pattern without trailing `$`

I could use `TODO(: [^\n\r]*)?` at the end of the regex, however, it also matches `TODO Something` and highlights the `TODO` word which I don't want as I consider it a style bug / typo. I want to be able to highlight only either `TODO$` or `TODO: .*$`, nothing else.

---

I still don’t really understand why `$` is not treated as _end of line_. I get it when `Treat indented text on the following lines as part of the same TODO` is enabled: then you need to switch the multiline matching on, however, if that setting is disabled, IMHO you should match all comments line by line. :thinking:

**Comment by jansorg** [comment]
Thanks for your update.
I'll think about if there's anything I can do about it.

Currently, most of the behavior you're observing is implemented by the IDE's platform code and I happen to use it. That platform TODO matching doesn't allow to customize the end-of-line matching.
The Shell Script plugin is different, because it's only providing basic comment support, which has a different code path in the IDE.

**Comment by tukusejssirs** [comment]
I have devoted some time to this and came up with the following two patterns instead of the first pattern in my list.

```regexp
# Main TODO pattern, but it does not match `[//]: # (TODO: Test comment)` used in Markdown
(?<=(/?[/*]| ?#|@remarks|rem|REM|::|--|;) )TODO(: [^\n\r]*|$)

# Markdown TODO pattern, but it covers much more than that
TODO(: [^\n\r]*|$)
```

There is still one issue: neither pattern matches `* TODO` in a multi-line comment in Java/JS/TS.

I don’t mind to have multiple TODO patterns, as long as it works as I want. :wink:

Here is how it works (no other TODO pattern besides those two have been defined):

![image](https://github.com/user-attachments/assets/a618ac29-fc06-4537-bae4-c79fc5f89529)

I forgot to add HTML and CSS snippets. `<!-- TODO -->` and `/* TODO */` does not work, but otherwise they’re fine.

![image](https://github.com/user-attachments/assets/13889b17-9c5e-4852-b59c-d999fc53bdb1)

**Comment by tukusejssirs** [comment]
The following pattern matches `/* TODO */`, but it does not work for the other two missed patterns.

```regex
(?<=(/?[/*]| ?#|@remarks|rem|REM|::|--|;) )TODO(: [^\n\r]*| (-->|\*)|$)
```

The strange thing is that this pattern matches both `<!-- TODO -->` and `* TODO` on [Regex101.com with the Java 8 flavour](https://regex101.com/r/uvmaqR/1).

**Comment by tukusejssirs** [comment]
Here is a single pattern which covers both my previously suggested patterns.

```regex
(?<=(/?[/*]| ?#|@remarks|rem|REM|::|--|;) |)TODO(: [^\n\r]*| (?=-->|\*)|$)
```

However, I have noticed something else what is different: now the patterns does not match comment characters (e.g. `//` or `#` or `/*` or `*`). Therefore, it is hard to match the TODO comments precisely.

**Comment by tukusejssirs** [comment]
I have found out that `(?<=\* )TODO[\n\r]` matches `* TODO`, but `(?<=\* )TODO$` does not. I presume that this is because of the multiline pattern matching.

Therefore, the following pattern works for everything but `<!-- TODO -->`:

```regex
(?<=(/?[/*]|\*| ?#|@remarks|rem|REM|::|--|;) |)TODO(: [^\n\r]*| (?=-->|\*)|[\n\r])
```

**Comment by tukusejssirs** [comment]
I have noticed that `[ ]TODO[ ]` is matching `<!-- TODO -->` which means that `<!--` and `-->` are not part of the TODO comment string on which the regex pattern is used. Note that `[ ]TODO[ ]` also matches `TODO` in `TODO Something to do` which I don’t want.

I understand that this is not an issue of BSP, however, it limits the pattern matching. I either use `[ ]TODO[ ]` which potentially matches more than I want, or not use it. :thinking:
